ipn/ipnlocal/cert.go

```diff
@@ -122,10 +124,78 @@ func shouldStartDomainRenewal(dir, domain string, future time.Time) bool {
 		return false
 	}
 	lastRenewCheck[domain] = now
-	_, ok := getCertPEMCached(dir, domain, future)
+	_, ok := b.getCertPEMCached(dir, domain, future)
 	return !ok
 }
 
+// certStore provides a way to perist and retrieve TLS certificates.
+// As of 2023-02-01, we use store certs in directories on disk everywhere
+// except on Kubernetes, where we use the state store.
+type certStore interface {
+	Read(domain string, now time.Time) (*TLSCertKeyPair, error)
+	WriteCert(domain string, cert []byte) error
+	WriteKey(domain string, key []byte) error
+}
+
+func (b *LocalBackend) getCertStore(dir string) certStore {
+	if hostinfo.GetEnvType() == hostinfo.Kubernetes {
+		return certStateStore{b.store}
+	}
+	return certFileStore(dir)
+}
```

Comment:
Not a big fan of this. Too magic and limiting... now people can't store certs on disk (some persistent volume?) like they could've before if so configured.

I'd rather have the wiring be explicit.

Reply:
We ignore the `StateDir` if a `KubeSecret` is specified. Added a check to make sure the underlying store is k8s secrets.
```
	args := []string{"--socket=" + cfg.Socket}
	switch {
	case cfg.InKubernetes && cfg.KubeSecret != "":
		args = append(args, "--state=kube:"+cfg.KubeSecret, "--statedir=/tmp")
	case cfg.StateDir != "":
		args = append(args, "--statedir="+cfg.StateDir)
	default:
		args = append(args, "--state=mem:", "--statedir=/tmp")
	}
```

Reply:
would you prefer an envknob? or another flag to tailscaled?

Reply:
I also don't think we should do this:

```go
	case cfg.InKubernetes && cfg.KubeSecret != "":
		args = append(args, "--state=kube:"+cfg.KubeSecret, "--statedir=/tmp")
```

... if StateDir is set to something else. We shouldn't magically overwrite/ignore their old value.

A default is fine.

Reply:
Fixed the StateDir handling in containerboot and added another check to use the kubestore only if the statedir is `/tmp`

Reply:
PTAL